Code/Tools/AssetProcessor/Platform/Windows/native/FileWatcher/FileWatcher_win.cpp

```diff
@@ -6,125 +6,149 @@
  *
  */
 
+#include <AzCore/std/tuple.h>
+#include <AzCore/std/utils.h>
 #include <native/FileWatcher/FileWatcher.h>
+#include <native/FileWatcher/FileWatcher_platform.h>
 
-#include <AzCore/PlatformIncl.h>
-
-struct FolderRootWatch::PlatformImplementation
-{
-    PlatformImplementation() : m_directoryHandle(nullptr), m_ioHandle(nullptr) { }
-    HANDLE m_directoryHandle;
-    HANDLE m_ioHandle;
-};
-
-//////////////////////////////////////////////////////////////////////////////
-/// FolderWatchRoot
-FolderRootWatch::FolderRootWatch(const QString rootFolder)
-    : m_root(rootFolder)
-    , m_shutdownThreadSignal(false)
-    , m_fileWatcher(nullptr)
-    , m_platformImpl(new PlatformImplementation())
+bool FileWatcher::PlatformStart()
 {
+    m_shutdownThreadSignal = false;
+
+    bool allSucceeded = true;
+    for (const auto& [directory, recursive] : m_folderWatchRoots)
+    {
+        if (QDir(directory).exists())
+        {
+            allSucceeded &= m_platformImpl->AddWatchFolder(directory, recursive);
+        }
+    }
+    return allSucceeded;
 }
 
-FolderRootWatch::~FolderRootWatch()
+bool FileWatcher::PlatformImplementation::AddWatchFolder(QString root, bool recursive)
 {
-    // Destructor is required in here since this file contains the definition of struct PlatformImplementation
-    Stop();
+    HandleUniquePtr directoryHandle{::CreateFileW(
+        root.toStdWString().data(),
+        FILE_LIST_DIRECTORY,
+        FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE,
+        nullptr,
+        OPEN_EXISTING,
+        FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED,
+        nullptr
+    )};
+
+    if (directoryHandle.get() == INVALID_HANDLE_VALUE)
+    {
+        AZ_Warning("FileWatcher", false, "Failed to start watching %s", root.toUtf8().constData());
+        return false;
+    }
 
-    delete m_platformImpl;
-}
+    // Associate this file handle with our existing io completion port handle
+    if (!::CreateIoCompletionPort(directoryHandle.get(), m_ioHandle.get(), 1, 1))
+    {
+        return false;
+    }
 
-bool FolderRootWatch::Start()
-{
-    m_platformImpl->m_directoryHandle = ::CreateFileW(m_root.toStdWString().data(), FILE_LIST_DIRECTORY, FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE, nullptr, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OVERLAPPED, nullptr);
+    auto id = AZStd::make_unique<OVERLAPPED>();
+    auto* idp = id.get();
+    const auto& [folderWatch, inserted] = m_folderRootWatches.emplace(AZStd::piecewise_construct, AZStd::forward_as_tuple(idp),
```

Comment:
Stylistic but I prefer the `try_emplace` interface to avoid needing to juggle piecewise construct and tuple forwarding and such